src/app/ipfs-cache.ts

```diff
@@ -78,13 +143,35 @@ export class IPFSCache {
    * @param      {Function}            callback  callback function, that is called
    *                                             when the transaction is finished
    */
-  _withIDBStore(type: IDBTransactionMode, callback: ((store: IDBObjectStore) => void)): Promise<void> {
-    return this._dbp.then((db) => new Promise<void>((resolve, reject) => {
-      const transaction = db.transaction(this.storeName, type);
-      transaction.oncomplete = () => resolve();
-      transaction.onabort = transaction.onerror = () => reject(transaction.error);
-      callback(transaction.objectStore(this.storeName));
-    }));
+  async execute(
+    type: IDBTransactionMode,
+    actionCallback: (store: IDBObjectStore) => any,
+    disableAbort = false,
+  ): Promise<any> {
+    const db = await this.getOpenedDb();
+    const transaction = db.transaction(this.storeName, type);
+    let request: any;
+
+    return new Promise((resolve, reject) => {
+      const handleError = (errorType: string): void => {
+        const isQuota = IPFSCache.isQuotaExceeded(transaction.error);
+        reject(new Error(`[ipfs-cache] - ${errorType} - ${transaction.error} - ${isQuota ? '(quota exceeded)' : ''}`));
+      };
+
+      transaction.oncomplete = (): void => resolve(request?.result);
```

Comment:
where is the `request` Object actually populated? Looks to me like we would always resolve to undefined ...

Reply:
nvm I just saw `request = actionCallback(transaction.objectStore(this.storeName));` at the end of this function.

The structure is really confusing, can we put the part from Line 173 inside the oncomplete function? Or even better first await transaction, make the code as linear as possible.

Reply:
We cannot move it into the oncomplete, because the oncomplete is the callback handle for the request.